workspaceitem-data-metadata.md

```diff
@@ -0,0 +1,297 @@
+* language: the iso code associated with the textual value if any
+
+## Patch operations
+The PATCH method expects a JSON body according to the [JSON Patch specification RFC6902](https://tools.ietf.org/html/rfc6902)
+
+### Add
+To add a new value to an **existent metadata** the client must send a JSON Patch ADD operation as follow
+
+`curl --data '{[ { "op": "add", "path": "/sections/<:name-of-the-form>/<:metadata>/-", "value": {value: "...", authority: "...", confidence: 600, language: ".."}}]}' -X PATCH ${dspace7-url}/api/submission/workspaceitems/<:id>`
+
+it is also possible to insert the new metadata in a specific position
+
+`curl --data '{[ { "op": "add", "path": "/sections/<:name-of-the-form>/<:metadata>/<:idx-zero-based>", "value": {value: "...", authority: "...", confidence: 600, language: ".."}}]}' -X PATCH ${dspace7-url}/api/submission/workspaceitems/<:id>`
+
+for example, starting with the following document  
+```json
+{
+	id: 1,
+	type: "workspaceitem",
+	sections:
+	{
+		"traditional-page1":
+		{
+		  "dc.title" : [{value: "Sample Submission Item", language: "en"}],
+		  "dc.contributor.author" : [
+		  	 		{value: "Bollini, Andrea", authority: "rp00001", confidence: 600}
+		  ]
+		},
+		...
+	},
+	...
+}	
+```
+
+the following request 
+`curl --data '{[ { "op": "add", "path": "/sections/traditional-page1/dc.contributor.author/-", "value": {value: "Another, Author"}]}' -X PATCH ${dspace7-url}/api/submission/workspaceitems/1`
+
+will result in
+```json
+{
+	id: 1,
+	type: "workspaceitem",
+	sections:
+	{
+		"traditional-page1":
+		{
+		  "dc.title" : [{value: "Sample Submission Item", language: "en"}],
+		  "dc.contributor.author" : [
+		  	 		{value: "Bollini, Andrea", authority: "rp00001", confidence: 600},
+		  	 		{value: "Another, Author", confidence: -1}
+		  ]
+		},
+		...
+	},
+	...
+}	
+```
+
+and an additional call as follow
+`curl --data '{[ { "op": "add", "path": "/sections/traditional-page1/dc.contributor.author/1", "value": {value: "Insert, Test", authority: "rp00002", confidence: 600}]}' -X PATCH ${dspace7-url}/api/submission/workspaceitems/1`
+
+will get the final document
+```json
+{
+	id: 1,
+	type: "workspaceitem",
+	sections:
+	{
+		"traditional-page1":
+		{
+		  "dc.title" : [{value: "Sample Submission Item", language: "en"}],
+		  "dc.contributor.author" : [
+		  	 		{value: "Bollini, Andrea", authority: "rp00001", confidence: 600},
+		  	 		{value: "Insert, Test", authority: "rp00002", confidence: 600},
+		  	 		{value: "Another, Author", confidence: -1}
+		  ]
+		},
+		...
+	},
+	...
+}	
+```
+
+Please note that according to the [JSON Patch specification RFC6902](https://tools.ietf.org/html/rfc6902) to initialize a new metadata in the section the add operation must receive an array of values and it is not possible to add a single value to the not yet initialized "/sections/<:name-of-the-form>/<:metadata>/-" path.
+For example the following call will add the first keyword to our previous example
```

Comment:
Can't we pre-initialise all allowed metadata fields with an empty array? Than it is immediately clear for a client to which metadata fields he can add values.

Reply:
I suggest to start following as much as possible the json patch specifications and only apply "hacks or customization" if we found them really needed by angular.
For instance if we pre-initialise the metadata what should be the behavior of a patch delete operation over the `/sections/<:name-of-the-form>/<:metadata>` The client should expect to have the metadata completely removed from the json response... should we deny this request? the client should use an *add* operation with an empty array to cleanup the metadata values but this is allowed but counter intuitive...

The client known which metadata can be added because the workspaceitem links the submissionDefinition that it uses.

Reply:
Ok I understand